[MO] Fix fp16 in shapeof subgraphs

pavel-esir opened the pull request:
Description: For some models `float16` precision in shape subgraphs is not sufficient for correct shape calculation - either shape value exceeds max(float16) or we get inaccurate scales values for operations like `Interpolate`.

In order to unblock such models decided to keep shape calculation in FP32 even if such models are being converted with command `--data_type=FP16`. To keep shapes in FP32 the following changes have been made:

1. added transformation `MarkNodesWithShapeValues` that goes up from ports of operations that accept shape values (Interpolate:1, Interpolate:2, Reshape:1, etc.) and marks such nodes with `'returns_shape_value'` attribute 
2. also the transformation below marks Const nodes carrying shape values with the attribute `'correct_data_type'` to prevent converting them to FP16 in `convert_data_type.py`
3. moved `ChangeCastOutputType` to the end of the back phase in order to cover all ShapeOf nodes that could've been inserted during common pipeline transformations
4.  in `ChangeCastOutputType` if Cast has attr `returns_shape_value==True` then keep return type float32 even if `['cmd_params'].data_type == 'FP16'`

Tickets: 42771, 44587

Code:
* [x]  Transformation tests
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names


Validation:
* [x]  Unit tests
* [x]  Framework operation tests, no new operations were added
* [x]  Model Optimizer IR Reader check: done

Documentation:
* [x]  Supported frameworks operations list: N/A
* [x]  Guide on how to convert the **public** model: N/A
* [x]  User guide update: N/A

pavel-esir commented on model-optimizer/extensions/middle/DecomposeBidirectionalRNNSequence.py:
```diff
@@ -60,8 +61,8 @@ def split_helper(node: Node, index: int, direction: str, axis: int = 0):
         return Op._create_data_node(
             node.graph,
             name=node.name + '/SplittedBiLSTM/{}/'.format(direction),
-            attrs={'value': np.take(node.value, [index], axis),
```
Agree. Will do this.

pavel-esir replied:
reverted

pavel-esir commented on model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:
```diff
@@ -72,10 +72,11 @@ def bfs(self, start_nodes: list, visited: set, condition: callable = None, forwa
         :param condition: function getting a Node as input and returning whether the node should be included into the
         result or not. If the value is None then the node is added unconditionally.
         :param forward: boolean flag specifying the traverse direction
+        :param include_both_directions: if True return both in and out nodes, forward param is ignored in that case
```
Agree. Wil do this.

pavel-esir replied:
done

pavel-esir commented on model-optimizer/extensions/middle/ReverseV2ToReverseSequence.py:
```diff
@@ -61,7 +62,7 @@ def replace_pattern(self, graph: Graph, match: dict):
         rename_node(reverse, reverse_name + '/to_delete')
         # 2. Create new ReverseSequence node and reconnect all inputs/outputs to it
-        reverse_sequence = create_op_node_with_second_input(graph, ReverseSequence, seq_lengths,
+        reverse_sequence = create_op_node_with_second_input(graph, ReverseSequence, float32_array(seq_lengths),
```
done

pavel-esir commented on model-optimizer/extensions/middle/MarkAndChangeDataTypeInShapeOfSubgraphs.py:
```diff
@@ -0,0 +1,48 @@
+
+
+class MarkShapeOfSubgraphDataType(MiddleReplacementPattern):
+    enabled = True
```
missed this comment will update after tests

pavel-esir replied:
Placed right before layout permutation. Added comment why we place the transformation at the end of middle

jane-intel commented on model-optimizer/extensions/middle/ChangeCastOutputType.py:
```diff
@@ -0,0 +1,55 @@
+            if node.dst_type == np.float32 and ir_data_type == np.float16 and not node.has_and_set('in_shape_subgraph'):
+                log.warning('Change data type from {} to {} for node {}'.format(node.dst_type, ir_data_type, node.name))
+                node.dst_type = ir_data_type
+            elif node.has_and_set('in_shape_subgraph') and node.dst_type == np.float16:
```
do we really need to introduce such a behavior? it may result in type infer collisions.

pavel-esir replied:
Yes, we really need such behavior. Yes, you are right that this can cause type infer collisions. But risk of that is minimal. 

I'm not much aware about type checks in ngraph, i know only that eltwise, concat ops must not have inputs with different types. There can be collisions but only if we want to Concat, Eltwise etc. shape values with values from Parameter input. If we are willing to do that there is something fundamentaly wrong with the original model and shape collisions in that case is a bigger problem than type mismatch

achetver reviewed (approved):
Looks good to me

pavel-esir commented:
changes are not completed yet, please do not review

pavel-esir commented on model-optimizer/extensions/back/MarkNodesWithShapeValues.py:
```diff
@@ -0,0 +1,82 @@
+
+    @staticmethod
+    def get_operations_with_shape_inputs():
+        return {
```
consider adding `StridedSlice` and `Range` to this dict

pavel-esir replied:
done

pavel-esir commented:
Updated list of operations with shape inputs. Added new operations: `VariadicSplit:2` with split_lengths, `OneHot:1` with depth input and many others.

lazarevevgeny commented on model-optimizer/extensions/back/ChangeCastOutputType.py:
```diff
@@ -0,0 +1,55 @@
+            if node.dst_type == np.float32 and ir_data_type == np.float16 and not node.has_and_set('returns_shape_value'):
+                log.warning('Change data type from {} to {} for node {}'.format(node.dst_type, ir_data_type, node.name))
+                node.dst_type = ir_data_type
+            elif node.has_and_set('returns_shape_value') and node.dst_type == np.float16:
+                # return back FP32 for all Convert nodes with shape values
```
How it can happen that the "Cast" node has `dst_type` equal to fp16?

pavel-esir replied:
There was a lot of situations when transformations inserted either FP16 or FP64 Consts I revised transformations but leaved a debug warning for the future if someone will introduce such new transformations. Also potentially this could happen if the original model had Cast to fp16.

lazarevevgeny commented on model-optimizer/extensions/back/MarkNodesWithShapeValues.py:
```diff
@@ -0,0 +1,88 @@
+
+        start_nodes = []
+        for node in nodes_with_shape_inputs:
+            start_ports = [x for x in shape_input_ops_map[node.type] if not node.in_port(x).disconnected()]
+            start_nodes.extend([node.in_port(port_idx).get_source().node for port_idx in start_ports])
```
Change these two lines to the following: 
```python
start_nodes.extend([node.in_port(port_idx).get_source().node for port_idx in shape_input_ops_map[node.type] if node.is_in_port_connected(port_idx)])
```

pavel-esir replied:
agree. done

lazarevevgeny commented on model-optimizer/extensions/back/MarkNodesWithShapeValues.py:
```diff
@@ -0,0 +1,88 @@
+    def find_and_replace_pattern(self, graph: Graph):
+        shape_input_ops_map = self.get_operations_with_shape_inputs()
+        nodes_with_shape_inputs = []
+        for op_type in shape_input_ops_map:
+            nodes_with_shape_inputs.extend(graph.get_op_nodes(type=op_type))
```
This approach is a bit non-performant. We prepare list of all ops in the graph many times (equal to the number of keys in the `get_operations_with_shape_inputs()`). I suggest to do the following.

```python
op_types_with_shape_inputs = self.get_operations_with_shape_inputs()
for node in graph.get_op_nodes():
    if node.soft_get('type') in op_types_with_shape_inputs:
        nodes_with_shape_inputs.append(node)
```

This way we prepare a list of graph nodes only ones and check if the node type is in the set only once for each node.

pavel-esir replied:
Agreed and changed according to your proposal

lazarevevgeny commented on model-optimizer/extensions/back/MarkNodesWithShapeValues.py:
```diff
@@ -0,0 +1,88 @@
+        nodes_with_shape_values = MarkSubGraphsWithCorrectLayout.bfs(start_nodes, set(), condition, forward=False)
+        for node in nodes_with_shape_values:
+            node['returns_shape_value'] = True
+            if node.type == 'Const':
```
`if node.soft_get('type') == 'Const':`

pavel-esir replied:
done

lazarevevgeny commented on model-optimizer/extensions/back/MarkNodesWithShapeValues.py:
```diff
@@ -0,0 +1,88 @@
+    """
+    This replacer marks op nodes in ShapeOf subgraphs with 'returns_shape_value' bool attribute and
+    data nodes of float32 constants with 'correct_data_type' attribute.
+    So that float Consts and Cast_to float will be kept in FP32 even if cmd_param --data_type=FP16 was specified.
```
"Cast_to" ? Please, change.

"cmd_param"? Please, change

"was" -> "is"

pavel-esir replied:
fixed

lazarevevgeny commented on model-optimizer/extensions/back/MarkNodesWithShapeValues.py:
```diff
@@ -0,0 +1,88 @@
+
+class MarkNodesWithShapeValues(BackReplacementPattern):
+    """
+    This replacer marks op nodes in ShapeOf subgraphs with 'returns_shape_value' bool attribute and
```
"replacer"->"transformation"

pavel-esir replied:
fixed

lazarevevgeny commented on model-optimizer/extensions/back/MarkNodesWithShapeValues.py:
```diff
@@ -0,0 +1,88 @@
+        condition = lambda node: node.soft_get('type') != 'ShapeOf'
+        nodes_with_shape_values = MarkSubGraphsWithCorrectLayout.bfs(start_nodes, set(), condition, forward=False)
+        for node in nodes_with_shape_values:
+            node['returns_shape_value'] = True
```
Probably, we need to set  `'returns_shape_value'` to `False` for all nodes in the graph first. Like this:
`nx.set_node_attributes(G=graph, name='returns_shape_value', values=False)`

pavel-esir replied:
We have wonderful methods like `has_and_set` or `soft_get` and use them quite often. I don't see any necessity for setting this attribute for every singe node in the graph.

lazarevevgeny commented on model-optimizer/extensions/back/MarkNodesWithShapeValues.py:
```diff
@@ -0,0 +1,88 @@
+            if node.type == 'Const':
+                if node.value.dtype == np.float32:
+                    node.out_node(0)['correct_data_type'] = True
+                elif node.value.dtype in [np.float16, np.float64]:
+                    log.debug('Const nodes {} with shape values have {} type'.format(node.soft_get('name', node.id),
+                                                                                     node.value.dtype))
```
Why do we need this check and log?

pavel-esir replied:
If float16 or float64 Const suddenly appear in shape flow it is an untypical and suspicious event indicating that either the original model had such Consts or in some MO transformation we inserted such Consts. I would prefer not to change by force data_type to **FP32** for the following reasons:
1. maybe model developer did that for a reason.
2. if we inserted in some transformations such Consts then it's better to revise out transformations and this log message will unhide such transformations.

lazarevevgeny replied:
I am sure that nobody will find this message in HUGE debug log :) Also, we have FP16 models from ONNX, so there is a high chance that FP16 appear in a shape sub-graph for them. Anyway, resolving the comment

lazarevevgeny commented on model-optimizer/extensions/back/MarkNodesWithShapeValues.py:
```diff
@@ -0,0 +1,88 @@
+    or because of FP16 lower precession shape inference is incorrect on some nodes (e.g. if Interpolate in scales mode
+    accepts values from ShapeOf subgraph).
+
+    This transformation must be run after shape inference and after all transformations that insert/modify
```
"must be run " -> "should be executed"

lazarevevgeny replied:
that -> which

pavel-esir replied:
done

pavel-esir commented on model-optimizer/extensions/back/ChangeCastOutputType.py:
```diff
@@ -0,0 +1,55 @@
+class ChangeCastOutputType(BackReplacementPattern):
+    """
+    Change the Cast to fp64 to fp32 since not all plugins support fp64 data type.
+    Change the Cast to fp32 to fp16 when generating IR for fp16.
```
updated

pavel-esir commented on model-optimizer/extensions/back/ChangeCastOutputType.py:
```diff
@@ -0,0 +1,55 @@
+    def find_and_replace_pattern(self, graph: Graph):
+        for node in graph.get_op_nodes(op='Cast'):
+            if node.dst_type == np.float64:
+                log.warning('Change data type from {} to {} for node {}'.format(node.dst_type, np.float32, node.name))
```
in this function we use node name just once

lazarevevgeny replied:
Node may have no `name` attribute. So we need to use `soft_get` method.

lazarevevgeny commented:
@pavel-esir , please, resolve merge conflicts.

pavel-esir commented:
resolved all conflicts. @lazarevevgeny @jane-intel please review
